// @raycast/eslint-plugin/dist/rules/prefer-ellipsis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("@typescript-eslint/utils");
const utils_2 = require("../utils");
exports.default = (0, utils_2.createRule)({
    create: (context) => {
        return {
            JSXOpeningElement: (node) => {
                if ((0, utils_2.isActionComponent)(node.name)) {
                    const titleAttribute = node.attributes.find((attribute) => {
                        return (attribute.type === "JSXAttribute" &&
                            attribute.name.name === "title");
                    });
                    if (titleAttribute) {
                        const value = titleAttribute.type === "JSXAttribute" && titleAttribute.value;
                        if (value) {
                            if (value.type === utils_1.AST_NODE_TYPES.Literal &&
                                typeof value.value === "string") {
                                const originalTitle = value.value;
                                const formattedTitle = originalTitle.replace(/\.\.\./g, "…");
                                if (originalTitle !== formattedTitle) {
                                    context.report({
                                        node: value,
                                        messageId: "isNotEllipsisCharacter",
                                        fix: (fixer) => fixer.replaceText(value, `"${formattedTitle}"`),
                                    });
                                }
                            }
                            if (value.type === utils_1.AST_NODE_TYPES.JSXExpressionContainer) {
                                const expression = value.expression;
                                if (expression.type === utils_1.AST_NODE_TYPES.Literal &&
                                    expression.value &&
                                    typeof expression.value === "string") {
                                    const formattedTitle = expression.value.replace(/\.\.\./g, "…");
                                    if (expression.value !== formattedTitle) {
                                        context.report({
                                            node: expression,
                                            messageId: "isNotEllipsisCharacter",
                                            fix: (fixer) => fixer.replaceText(expression, `"${formattedTitle}"`),
                                        });
                                    }
                                }
                                if (expression.type === utils_1.AST_NODE_TYPES.ConditionalExpression) {
                                    const consequent = expression.consequent;
                                    const alternate = expression.alternate;
                                    if (consequent.type === utils_1.AST_NODE_TYPES.Literal &&
                                        consequent.value &&
                                        typeof consequent.value === "string") {
                                        const formattedTitle = consequent.value.replace(/\.\.\./g, "…");
                                        if (consequent.value !== formattedTitle) {
                                            context.report({
                                                node: consequent,
                                                messageId: "isNotEllipsisCharacter",
                                                fix: (fixer) => fixer.replaceText(consequent, `"${formattedTitle}"`),
                                            });
                                        }
                                    }
                                    if (alternate.type === utils_1.AST_NODE_TYPES.Literal &&
                                        alternate.value &&
                                        typeof alternate.value === "string") {
                                        const formattedTitle = alternate.value.replace(/\.\.\./g, "…");
                                        if (alternate.value !== formattedTitle) {
                                            context.report({
                                                node: alternate,
                                                messageId: "isNotEllipsisCharacter",
                                                fix: (fixer) => fixer.replaceText(alternate, `"${formattedTitle}"`),
                                            });
                                        }
                                    }
                                }
                                if (expression.type === utils_1.AST_NODE_TYPES.TemplateLiteral) {
                                    const quasis = expression.quasis;
                                    var hasQuasiWithoutTitleCase = false;
                                    quasis.forEach((quasi) => {
                                        if (quasi.type === utils_1.AST_NODE_TYPES.TemplateElement &&
                                            quasi.value &&
                                            typeof quasi.value.raw === "string") {
                                            const formattedTitle = quasi.value.raw.replace(/\.\.\./g, "…");
                                            if (quasi.value.raw !== formattedTitle) {
                                                hasQuasiWithoutTitleCase = true;
                                            }
                                        }
                                    });
                                    if (hasQuasiWithoutTitleCase) {
                                        return context.report({
                                            node: expression,
                                            messageId: "isNotEllipsisCharacter",
                                        });
                                    }
                                }
                            }
                        }
                    }
                }
            },
        };
    },
    name: "prefer-ellipsis",
    meta: {
        fixable: "code",
        messages: {
            isNotEllipsisCharacter: "Prefer `…` to `...`",
        },
        type: "suggestion",
        docs: {
            description: "Prefer Ellipsis Character",
        },
        schema: [],
    },
    defaultOptions: [{}],
});
